peekingduck/pipeline/nodes/model/jde_mot/jde_files/utils/utils.py

```diff
@@ -0,0 +1,256 @@
+    of bottom left and top right respectively.
+
+    Args:
+        x (torch.Tensor): Tensor with bounding boxes of format
+            `(center x, center y, width, height)`.
+
+    Returns:
+        torch.Tensor: Tensor with bounding boxes of format
+            `(top left x, top left y, bottom right x, bottom right y)`.
+    """
+    y = torch.zeros_like(x) if x.dtype is torch.float32 else np.zeros_like(x)
+    y[:, 0] = x[:, 0] - x[:, 2] / 2  # Bottom left x
+    y[:, 1] = x[:, 1] - x[:, 3] / 2  # Bottom left y
+    y[:, 2] = x[:, 0] + x[:, 2] / 2  # Top right x
+    y[:, 3] = x[:, 1] + x[:, 3] / 2  # Top right y
+    return y
+
+
+def generate_anchor(
+    nGh: int, nGw: int, anchor_wh: torch.Tensor, device: torch.device
+) -> torch.Tensor:
+    """Generate anchor bounding box for prediction.
+
+    Args:
+        nGh (int): Number of grid height.
+        nGw (int): Number of grid width.
+        anchor_wh (torch.Tensor): Anchor for bounding boc width, height.
+        device (torch.device): Device type being used. "cuda" or "cpu".
+
+    Returns:
+        torch.Tensor: Prediction bounding box anchors.
+    """
+    nA = len(anchor_wh)
+    yy, xx = torch.meshgrid(torch.arange(nGh), torch.arange(nGw))
+    xx = xx.to(device)
+    yy = yy.to(device)
+
+    mesh = torch.stack([xx, yy], dim=0)  # Shape 2, nGh, nGw
+    mesh = mesh.unsqueeze(0).repeat(nA, 1, 1, 1).float()  # Shape nA x 2 x nGh x nGw
+    anchor_offset_mesh = (
+        anchor_wh.unsqueeze(-1).unsqueeze(-1).repeat(1, 1, nGh, nGw)
+    )  # Shape nA x 2 x nGh x nGw
+    anchor_mesh = torch.cat(
+        [mesh, anchor_offset_mesh], dim=1
+    )  # Shape nA x 4 x nGh x nGw
+    return anchor_mesh
+
+
+def decode_delta(delta: torch.Tensor, fg_anchor_list: torch.Tensor) -> torch.Tensor:
+    """Decode delta tensors.
+
+    Args:
+        delta (torch.Tensor): Detection prediction tensor.
+        fg_anchor_list (torch.Tensor): Anchors for bounding box.
+
+    Returns:
+        torch.Tensor: Prediction tensors.
+    """
+    px, py, pw, ph = (
+        fg_anchor_list[:, 0],
+        fg_anchor_list[:, 1],
+        fg_anchor_list[:, 2],
+        fg_anchor_list[:, 3],
+    )
+    dx, dy, dw, dh = delta[:, 0], delta[:, 1], delta[:, 2], delta[:, 3]
+    gx = pw * dx + px
+    gy = ph * dy + py
+    gw = pw * torch.exp(dw)
+    gh = ph * torch.exp(dh)
+    return torch.stack([gx, gy, gw, gh], dim=1)
+
+
+def decode_delta_map(
+    delta_map: torch.Tensor,
+    anchors: Union[torch.Tensor, torch.nn.Module],
+    device: torch.device,
+) -> torch.Tensor:
+    """Decode delta map.
+
+    Args:
+        delta_map (torch.Tensor): shape (nB, nA, nGh, nGw, 4).
+        anchors (torch.Tensor): shape (nA,4).
+
+    Returns:
+        torch.Tensor: Prediction map.
+    """
+    nB, nA, nGh, nGw, _ = delta_map.shape
+    anchor_mesh = generate_anchor(nGh, nGw, anchors, device)  # type: ignore
+    anchor_mesh = anchor_mesh.permute(
+        0, 2, 3, 1
+    ).contiguous()  # Shpae (nA x nGh x nGw) x 4
```

Comment:
Shpae -> Shape and avoid inline comment if it causes code to wrap

Reply:
Updated